stdlib/public/core/GraphemeIterator.swift

```diff
@@ -0,0 +1,341 @@
+    if let cache = cachedGBP, cache.0 == scalar {
+      return cache.1
+    }
+
+    let gbp = Unicode.GraphemeBreakProperty(from: scalar)
+    cachedGBP = (scalar, gbp)
+    return gbp
+  }
+}
+
+
+extension GraphemeIterator {
+  // Returns the stride of the next grapheme cluster at the previous boundary at
+  // offset.
+  mutating func next() -> Int {
+    inEmojiSequence = false
+    shouldBreakRI = false
+
+    while scalars.index(after: index) != scalars.endIndex {
+      let sc1 = scalars[index]
+      scalars.formIndex(after: &index)
+
+      let sc2 = scalars[index]
+
+      let gbp1 = getGraphemeBreakProperty(for: sc1)
+      let gbp2 = getGraphemeBreakProperty(for: sc2)
+
+      if shouldBreak(gbp1, between: gbp2, in: &self) {
+        return index._encodedOffset &- offset
+      }
+    }
+
+    // If we've reached the end of the buffer, then there's nothing else for
+    // us to do.
+    return scalars.index(after: index)._encodedOffset &- offset
+  }
+
+  // Returns the stride of the previous grapheme cluster at the current boundary
+  // at offset.
+  mutating func previous() -> Int {
+    scalars.formIndex(before: &index)
+
+    repeat {
+      let sc2 = scalars[index]
+      let sc2Idx = index
+
+      scalars.formIndex(before: &index)
+
+      let sc1 = scalars[index]
+
+      let gbp2 = getGraphemeBreakProperty(for: sc2)
+      let gbp1 = getGraphemeBreakProperty(for: sc1)
+
+      if shouldBreak(gbp1, between: gbp2, in: &self) {
+        return offset &- sc2Idx._encodedOffset
+      }
+    } while index != scalars.startIndex
+
+    // If we've reached the beginning of the buffer, then there's nothing else
+    // for us to do.
+    return offset &- index._encodedOffset
+  }
+}
+
+// The "algorithm" that determines whether or not we should break between
+// certain grapheme break properties.
+//
+// This is based off of the Unicode Annex #29 for [Grapheme Cluster Boundary
+// Rules](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundary_Rules).
+internal func shouldBreak(
+  _ x: Unicode.GraphemeBreakProperty,
+  between y: Unicode.GraphemeBreakProperty,
+  in iterator: inout GraphemeIterator
+) -> Bool {
+  var enterEmojiSequence = false
+
+  defer {
+    iterator.inEmojiSequence = enterEmojiSequence
+  }
+
+  switch (x, y) {
+
+  // GB3
+  case (.cr, .lf):
+    return false
+
+  // GB4
+  case (.control, _),
+       (.cr, _),
+       (.lf, _):
+    return true
+
+  // GB5
+  case (_, .control),
+       (_, .cr),
+       (_, .lf):
+    return true
+
+  // GB6
+  case (.l, .l),
+       (.l, .v),
+       (.l, .lv),
+       (.l, .lvt):
+    return false
+
+  // GB7
+  case (.lv, .v),
+       (.v, .v),
+       (.lv, .t),
+       (.v, .t):
+    return false
+
+  // GB8
+  case (.lvt, .t),
+       (.t, .t):
+    return false
+
+  // GB9 (partial GB11)
+  case (_, .extend),
+       (_, .zwj):
+    if iterator.inEmojiSequence ?? false || x == .extendedPictographic {
+      enterEmojiSequence = true
```

Comment:
This feels like a complicated cross-call convention, so hopefully we can clarify it somewhat.